Iterator string smarts

som-snytt opened the pull request:
Many iterators know if they are empty or not. But they can't tell us.

Also trailing iterator of a span is more reluctant to forward
hasNext.

sjrd commented:
So much code to put in a .js file for ... what benefit exactly?

som-snytt commented:
@sjrd Just for a break. I was waiting on someone else's commit at the office.

SethTisue commented:
as discussed at scala/bug#10892, I don't think we should make this change.

dwijnand commented:
Hey @som-snytt thanks for the path. I like how you're using stable/pure information to make toString as informative as possible.

However I'm not sure about this solution as the REPL output you put in scala/bug#10892 seems a bit.. schizophrenic.

In terms of change size, ignoring tests this is about a -25/+50 LOC change. Perhaps that's a bit on the heavy side.

What's your motivation (or thinking) for proposing this solution?

Oh and thanks for picking up my ticket so fast! <3

som-snytt committed:
Avoid extra hasNext in trailing

Once trailing was advanced, hasNext was
needlessly delegated to the underlying
iterator on every invocation.

som-snytt commented:
Thanks all for looking. It does look weird; it would make more sense as special cases in `stringOf` or a similar typeclass; the motivation is that these iterators know more about themselves than they let on, and someone learning about their behavior in REPL deserve to share that info. Not that I feel strongly about it.

I'll reopen with just the fix to the trailing iterator of a span result.

som-snytt committed:
Avoid side-effects in Iterator.toString

Don't probe hasNext in Iterator.toString.
